bus-mapping/src/exec_trace/exec_step.rs

```diff
@@ -0,0 +1,198 @@
+// Doc this
+
+use crate::evm::{
+    EvmWord, GlobalCounter, Instruction, MemoryAddress, ProgramCounter, StackAddress, MEM_ADDR_ZERO,
+};
+use crate::{
+    error::Error,
+    evm::opcodes::Opcode,
+    operation::{bus_mapping::BusMappingInstance, container::OperationContainer},
+};
+use alloc::collections::BTreeMap;
+use core::{convert::TryFrom, str::FromStr};
+use halo2::arithmetic::FieldExt;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// Doc
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct ExecutionStep {
+    memory: BTreeMap<MemoryAddress, EvmWord>,
+    stack: Vec<EvmWord>,
+    instruction: Instruction,
+    pc: ProgramCounter,
+    gc: GlobalCounter,
+    // Holds refs to the container with the related mem ops.
+    bus_mapping_instances: BusMappingInstance,
+}
+
+impl ExecutionStep {
+    /// Doc
+    pub fn new(
+        memory: BTreeMap<MemoryAddress, EvmWord>,
+        stack: Vec<EvmWord>,
+        instruction: Instruction,
+        pc: ProgramCounter,
+        gc: GlobalCounter,
+    ) -> Self {
+        ExecutionStep {
+            memory,
+            stack,
+            instruction,
+            pc,
+            gc,
+            bus_mapping_instances: BusMappingInstance::new(),
+        }
+    }
+
+    pub const fn memory(&self) -> &BTreeMap<MemoryAddress, EvmWord> {
+        &self.memory
+    }
+
+    pub const fn stack(&self) -> &Vec<EvmWord> {
+        &self.stack
+    }
+
+    pub fn stack_addr(&self) -> StackAddress {
+        // Stack has 1024 slots.
+        // First allocation slot for us in the stack is 1023.
+        StackAddress::from(1024 - self.stack.len())
+    }
+
+    pub fn memory_addr(&self) -> &MemoryAddress {
+        self.memory
+            .iter()
+            .last()
+            .map(|items| items.0)
+            .unwrap_or_else(|| &MEM_ADDR_ZERO)
+    }
+
+    pub const fn instruction(&self) -> &Instruction {
+        &self.instruction
+    }
+
+    pub const fn pc(&self) -> ProgramCounter {
+        self.pc
+    }
+
+    pub const fn gc(&self) -> GlobalCounter {
+        self.gc
+    }
+
+    pub fn set_gc(&mut self, gc: impl Into<GlobalCounter>) {
+        self.gc = gc.into()
+    }
+
+    pub const fn bus_mapping_instances(&self) -> &BusMappingInstance {
+        &self.bus_mapping_instances
+    }
+
+    pub fn bus_mapping_instances_mut(&mut self) -> &mut BusMappingInstance {
+        &mut self.bus_mapping_instances
+    }
+
+    // Returns the # operations added by the opcode
```

Comment:
Document the mutation of `container` as well.

```suggestion
    // Returns the # operations added by the opcode
```

Reply:
Not sure I follow. WDYM by documenting the mutation?